### PR TITLE
[NEW mrp_lock_lot]

### PR DESCRIPTION
**Pull request by alfredoavanzosc**
Este nuevo módulo es para cubrir los nuevos requerimientos de la tarea de OdooMRP DESA1. Bloqueo de lotes.

### CODE REVIEW

**Comment by clonedagain** [comment]
We have the same requirements and we developed a module which is very similar.
However we went a bit farther by setting a python constraint to refuse to finish stock moves using locked lots (except for inventories if I remember correctly).
Given this, our users found it better to use filters on the search view (enabled by default) to mask the locked lots instead of hard domains. This way it lets them disable the filter when they want: they can see the locked lots so they can know it's not an inventory problem, and they can record draft moves that can be executed when the lot is unlocked.
Another user wish is that all locking/unlocking be logged, possibly with a comment or reason, for audit purposes. This would be easily done by adding a chatter to the production lot.

**Comment by anajuaristi** [comment]
Hi @clonedagain 
I think we should combine efforts and made a single module. Never mind if we take yours or ours, question is taking the best from both and having a single one to mantain on future.

Doubs about our module:
.- I was thinking about if it would not be better using an active field to simulate lot blocking since that way we could have restriction by default for blocked/inactive lots so we will need less programming lines.
.- About using a default filter for searching I think is a good idea but.. we should have a filter (not by defautl) to allow user to look for inactive lots in puntual places. That is... we could use an inactive lot to be moved from one internal location to another and also to make inventory but never in movements related to an outgoing picking or related to manufacturing order. 
.- Chatter on lot is also a good Idea. 
.- We should also make a default filter on lots to see all (active and inactive) on tree. I would also show active field on lot's tree. 

Could you make a PR of your module in order to see if it would be easier to select what module take to be the base of the final one? 

@pedrobaeza Opinions about these new ideas?

**Comment by pedrobaeza** [comment]
OK for me to take the best approach. @clonedagain, please share your module (is for v7, right?) and we'll take a lot of it and share opinions here.

**Comment by clonedagain** [comment]
I've pushed our v7 module in ~~https://github.com/numerigraphe/numerigraphe-addons/commit/43a6d6c85fb3b127e07b475431e9ab3ed9ec00ab~~ https://github.com/numerigraphe/numerigraphe-addons/commit/957e2e69cf830d1bfa81a6db313beb27523b01ca (relicensed as AGPL and PEP8ified).
It depends on other private modules and the wording does not suite the context of OdooMRP, but you'll get the picture.

**Comment by clonedagain** [comment]
If you investigate our module, be sure to get the latest commit: https://github.com/numerigraphe/numerigraphe-addons/commit/957e2e69cf830d1bfa81a6db313beb27523b01ca .

The one I posted earlier has a nasty client bug in `search()` : no matter what, `search()` MUST NOT filter out records when `context['active_test'] is False`.
This is because when records are created by the client, it calls `search()` with the id it just created (for some esoteric reason I dare not imagine).
@lbellier just spend the day hunting the weird client bug it triggers.

We've also cleaned the code a bit more.

File: mrp_lock_lot/views/mrp_view.xml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+                This is the list of all the production lots you recorded. When
+                you select a lot, you can get the 
+                traceability of the products contained in lot. By default, the
+                list is filtered on the serial numbers that are available in
+                your warehouse but you can uncheck the 'Available' button to
+                get all the lots you produced, received or delivered to
+                customers.
+              </p>
+            </field>
+        </record>
```
Comment (clonedagain) [inline mrp_lock_lot/views/mrp_view.xml:6]: You should not completely replace the standard view, only inherit it, shouldn't you?

**Comment by alfredoavanzosc** [comment]
I've done it.

**Comment by anajuaristi** [comment]
@pedrobaeza 
Esto podriamos hacer merge? Sería urgente.

**Comment by pedrobaeza** [comment]
Tengo que revisar la propuesta de Lionel, pero lo tengo pendiente.

**Comment by anajuaristi** [comment]
Hello everybody:

After taking a look to V7 stock_prodlot_lock shared by @clonedagain, IMHO it doesn't fit our customers needs and... since It was built for V7, I think it also needs to be completly refactored to be adapted to the new VMS aproach, where lots are related to quants instead of movements.

Module aproach is not allowing confirming an stock move having related locked lot. 
Now... aproach should be, just don't allow to assing a quant on consumption moves having locked lot. It's totally different aproach.

On the other side, we should take in account --> Don't allow inactivating a lot, if exists a quant containing it in a movement with virtual destiny location (this is: consumption move)
By default lots will be created "inactive" way until a user activates them, so it will not be allowed to consume until user decides

@pedrobaeza I think we should review our own module and finish it with this aproach.

**Comment by clonedagain** [comment]
OK enlightening opinion.
When we move to v8 (soon hopefully) I think we'll be better off dropping 
our own module in favor of yours when it's done, unless @lbellier thinks 
otherwise.

File: mrp_lock_lot/README.rst [hunk 2]
```diff
@@ -0,0 +1,8 @@
+    This module performs the following:
```
Comment (pedrobaeza) [inline mrp_lock_lot/README.rst:1]: Poner título y descripción más "user-friendly"

File: mrp_lock_lot/i18n/es.po [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+#. module: mrp_lock_lot
+#: view:stock.production.lot:mrp_lock_lot.search_product_lot_filter_inh_locklot
+#: field:stock.production.lot,active:0
+msgid "Active"
+msgstr "Activo"
+
+#. module: mrp_lock_lot
+#: model:ir.model,name:mrp_lock_lot.model_stock_production_lot
+msgid "Lot/Serial"
```
Comment (pedrobaeza) [inline mrp_lock_lot/i18n/es.po:26]: Puedes eliminar esta frase y la siguiente del pot y el po, porque sobreescriben la estándar.

**Comment by pedrobaeza** [comment]
Para ser semánticamente correctos, deberíamos tener un campo boolean locked, y active debería ser un campo calculado que sea `not locked`. Después, esos campos que sólo se puedan modificar vía un botón "Bloquear", que primero compruebe si los quants de ese lote están reservados para algo, y si es así, que no permita bloquear.

Después, otra incongruencia que veo, es que no se modifica para nada el método quant_search que busca los quants a asignar por ejemplo cuando se pulsa el botón comprobar disponibilidad, para eliminar a aquellos que pertenezcan a un lote bloqueado.

Conforme he ido escribiendo esto, he visto que el campo active no haría falta con esto ni tampoco serviría para nada más, ya que el que un registro esté inactivo no va a inhibir el proceso de reserva de quants.

**Comment by alfredoavanzosc** [comment]
Sobre esto último que nos comenta Pedro. Yo en un principio definí un campo llamada "locked", luego Ana me dijo que de eso nada, que usaria un campo "active", que funcionalmente fuese como siempre.
Os quedaría muy agradecido si me volsieséis a escribir los requisitos, ya que no entiendo a Pedro cuando dice "active debería ser un campo calculado que sea not locked", y luego dice "he visto que el campo active no haría falta con esto ni tampoco serviría para nada más".
Entiendo que el módulo debería de quedar como al principio, desaparece el campo "active", le ponemos el campo "Locked", ponemos un botón de "Bloquear Lote", y hacemos la validación de que no existan quands. ¿es así?

**Comment by anajuaristi** [comment]
Si. Tiene razón @alfredoavanzosc en este caso.
inicialmente se hizo con el campo locked y luego se cambio a un active porque al estar inactivo ya desaparece de varias vistas.

el uno o el otro, son campos booleanos, y al final daría un poco lo mismo. 

Las tareas que hay que hacer (como si empezásemos de cero con esto, aunque habrá cosas ya hechas en el módulo)
1. Definir un campo locked de tipo boolean en los lotes (como al principio). POR DEFECTO TRUE 
    Definir un campo locked de tipo boolean en los quants --> Related a lote. POR DEFECTO FALSE, SOLO será TRUE si tiene un lote con campo locked a TRUE.  
2. Añadir un botón "Bloquear/desbloquear" en lotes 
Si un lote está bloqueado, se mostrará el botón desbloquear.
Si un lote está desbloqueado, se mostrará el botón bloquear. 
Al Bloquear un lote --> Verificar que no existe ningún movimiento de stock con ubicación destino de tipo virtual/empresa que tenga asignado un quant, que tenga dicho lote. Si existe --> Mostrar warning.
3. Hay que modificar la función "verificar disponiblidad" de todos los albaranes, cuyo tipo de albarán tenga ubicación destino virtual para que al realizar la asignación de quant, no tenga en cuenta quants "bloqueados" 
4. Hay que crear un módulo autoinstalable para cuando esté instalado "asignación quant manual" y este, se controle que al mostrar el wizard de asignación de lotes manual, no muestre quants "bloqueados"

En los puntos 3 y 4 indicación de @pedrobaeza -->  hay que heredar el método que obtiene el dominio para filtrar los quants (_get_quants_domain o algo así) y añadir que no tengan locked=True

Creo que esa sería la especificación correcta, si no me dejo nada.

**Comment by pedrobaeza** [comment]
Es más fácil que todo eso: hay que heredar el método que obtiene el dominio para filtrar los quants (_get_quants_domain o algo así) y añadir que no tengan locked=True

**Comment by anajuaristi** [comment]
Especificación final, validada por @pedrobaeza y por mi misma:

Las tareas que hay que hacer (como si empezásemos de cero con esto, aunque habrá cosas ya hechas en el módulo)
1. Definir un campo locked de tipo boolean en los lotes (como al principio). POR DEFECTO TRUE 
Definir un campo locked de tipo boolean en los quants --> Related a lote. POR DEFECTO FALSE, SOLO será TRUE si tiene un lote con campo locked a TRUE.
2. Añadir un botón "Bloquear/desbloquear" en lotes 
Si un lote está bloqueado, se mostrará el botón desbloquear.
Si un lote está desbloqueado, se mostrará el botón bloquear. 
Al Bloquear un lote --> Verificar que no existe ningún movimiento de stock con ubicación destino de tipo virtual/empresa que tenga asignado un quant, que tenga dicho lote. Si existe --> Mostrar warning.
3. Hay que modificar la función "verificar disponiblidad" de todos los albaranes, cuyo tipo de albarán tenga ubicación destino virtual para que al realizar la asignación de quant, no tenga en cuenta quants "bloqueados" 
4. Hay que crear un módulo autoinstalable para cuando esté instalado "asignación quant manual" y este, se controle que al mostrar el wizard de asignación de lotes manual, no muestre quants "bloqueados"

En los puntos 3 y 4 indicación de @pedrobaeza --> hay que heredar el método que obtiene el dominio para filtrar los quants (_get_quants_domain o algo así) y añadir que no tengan locked=True

**Pull request closed without merging**